core/src/avm2/names.rs

```diff
@@ -0,0 +1,309 @@
+//! AVM2 names & namespacing
+
+use crate::avm2::value::{abc_string, abc_string_option};
+use crate::avm2::{Avm2, Error};
+use gc_arena::Collect;
+use swf::avm2::types::{
+    AbcFile, Index, Multiname as AbcMultiname, Namespace as AbcNamespace,
+    NamespaceSet as AbcNamespaceSet,
+};
+
+/// Represents the name of a namespace.
+#[derive(Clone, Collect, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[collect(no_drop)]
+pub enum Namespace {
+    Namespace(String),
+    Package(String),
+    PackageInternal(String),
+    Protected(String),
+    Explicit(String),
+    StaticProtected(String),
+    Private(String),
```

Comment:
Same here, is it possible for these to be Cows? We seem to construct them frequently based on existing data

Reply:
Unfortunately, no. For various reasons, `Value` needs to be able to hold a `Namespace`, and if I stick a `Cow` in here then the lifetime I use has to leak out into everything that works with namespaces. I tried to work around this with `Namespace<'static>`, which I could fit into a `Value<'gc>`, but then implicit lifetimes broke in anything that needs to hold references to a `Value`, which turns out to be a lot of code (it's part of the definition of native methods/functions). This also broke in the same way with `Namespace<'gc>`, a lifetime bound which makes absolutely no sense.

Since there isn't a way to extinguish the borrowed half of the `Cow` when I specifically need to structurally require an owned value, I tried moving this up one notch by implementing a separate borrowed namespace type called `NS`. The idea was that you could hold a `Cow<'a, NS<'a>>`, and then have it return a `Namespace`. This also did not work, due to a chain of conflicting requirements:

 * In order to stick a type in a `Cow`, it needs to implement `ToOwned`.
 * In order to implement `ToOwned`, your type cannot be `Clone` (otherwise you get an auto implementation), which I was able to work around
 * In order to implement `ToOwned`, the type you do return must implement `Borrow`.
 * In order to implement `Borrow`, your owned type must return a reference to the borrowed type... or, `&'a NS<'a>`, which makes no sense for an enum being constructed on-stack.
 * Ergo, we can't make a pair of types that `Cow` can work with.

If you have any other suggestions about how to be able to pass borrowed-string namespaces, please let me know.

Reply:
I straight up replaced all the strings in AVM2 with `AvmString`, which should provide the same string deduplication opportunities as `Cow<'static, str>` does. I'm going to mark that as a resolution to this concern.